syntax = "proto3";

package grid;

// Grid Master service - coordinates agents and storage nodes
service GridMaster {
    // Agent lifecycle
    rpc RegisterAgent(AgentInfo) returns (RegistrationResponse);
    rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
    rpc UnregisterAgent(AgentId) returns (Empty);
    
    // Deployment control
    rpc SpawnStorageNode(SpawnRequest) returns (SpawnResponse);
    rpc StopStorageNode(StopRequest) returns (StopResponse);
    rpc GetStorageNodeStatus(NodeId) returns (NodeStatus);
    
    // Cluster info
    rpc ListAgents(Empty) returns (AgentList);
    rpc GetClusterStatus(Empty) returns (ClusterStatus);
}

// ===== Agent Registration =====

message AgentInfo {
    string agent_id = 1;
    string hostname = 2;
    string platform = 3;  // "docker", "k8s", "aws", "desktop"
    uint32 max_storage_nodes = 4;
    string version = 5;
    string agent_endpoint = 6;  // "hostname:port" for Agent gRPC server
}

message RegistrationResponse {
    bool success = 1;
    string master_version = 2;
    string error_message = 3;
}

// ===== Heartbeat =====

message HeartbeatRequest {
    string agent_id = 1;
    uint32 storage_node_count = 2;
    uint64 timestamp = 3;
}

message HeartbeatResponse {
    bool acknowledged = 1;
    uint64 timestamp = 2;
}

// ===== Storage Node Management =====

message SpawnRequest {
    string agent_id = 1;
    string storage_path = 2;
    uint64 memory_limit_mb = 3;
    uint32 port = 4;
}

message SpawnResponse {
    string node_id = 1;
    string endpoint = 2;  // host:port
    bool success = 3;
    string error_message = 4;
}

message StopRequest {
    string agent_id = 1;
    string node_id = 2;
}

message StopResponse {
    bool success = 1;
    string error_message = 2;
}

message NodeId {
    string node_id = 1;
}

message NodeStatus {
    string node_id = 1;
    string status = 2;  // "starting", "running", "stopping", "stopped", "failed"
    uint32 pid = 3;
    string endpoint = 4;
}

// ===== Cluster Info =====

message AgentList {
    repeated AgentRecord agents = 1;
}

message AgentRecord {
    string agent_id = 1;
    string hostname = 2;
    string platform = 3;
    string status = 4;  // "healthy", "degraded", "offline"
    uint32 max_storage_nodes = 5;
    uint32 current_storage_nodes = 6;
    uint64 last_heartbeat = 7;
    repeated NodeStatus storage_nodes = 8;
}

message ClusterStatus {
    uint32 total_agents = 1;
    uint32 healthy_agents = 2;
    uint32 total_storage_nodes = 3;
    uint32 running_storage_nodes = 4;
    string status = 5;  // "healthy", "degraded", "critical"
}

// ===== Common =====

message AgentId {
    string agent_id = 1;
}

message Empty {}
